<a href="https://colab.research.google.com/github/rlfwo93/machine-learning/blob/master/TF2_0_for_beginner.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tensorflow library import하기

In [ ]:
!pip install -q tensorflow-gpu==2.0.0-rc1
import tensorflow as tf

     |████████████████████████████████| 380.5MB 38kB/s 
     |████████████████████████████████| 501kB 35.2MB/s 
     |████████████████████████████████| 4.3MB 46.2MB/s 


# MNIST dataset 준비하기

In [ ]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train/255.0, x_test/255.0

## 왜 255.0으로 나눠주는 거지?
mnist.load_data()로 불러온 데이터들은 모두 정수형이다.
0 ~ 255사이의 값을 255.0으로 나눠주면 0 ~ 1사이 값으로 바뀌게된다. 이것을 정규화(normalization)이라고 한다.

## 왜 정규화가 필요해?
정규화는 feature들의 값을 왜곡하지 않고 scale을 공통적으로 변경시키는 것을 말하며 feature scaling이라고도 한다.

0 ~ 1 사이 값으로 데이터들이 정규화되면 Gradient Descent를 수행할 때 unnormalized data를 사용할 때보다 더 쉽고 빠르게 수행하여 optimum 지점에 가까워지게 된다.

**즉 학습이 잘되기 때문에 정규화를 하는 것이다.**

정규화의 일반적인 방법은 다음과 같다.
$$
    \frac{x-x_{min}}{x_{max}-x_{min}}
$$


# 모델을 만들자

In [ ]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

## model 구조
모델에 사용된 레이어들
* flatten : 1d로 평평하게 펴주는 레이어
* Dense : fully connected 레이어. 이전 레이어의 출력과 모든 뉴런들을 연결한다.
* Dropout : regularization을 위해 사용하는 Dropout, 특정 비율로 뉴런을 turn off해서 학습한다.

1. Flatten(input_shape=(28,28)) : 28x28 shape의 입력데이터를 28*28 x 1로 변경해준다.
2. Dense(128, activation='relu') : 128개의 뉴런을 가진 FC layer, 활성함수는 relu를 사용한다.
3. Dropout(0.2) : 20%의 뉴런을 drop한다.
4. Dense(10, activation='softmax') : 10개의 뉴런을 가진 FC layer, 활성함수는 softmax를 사용한다.

output layer의 뉴런이 10개이고 softmax를 활성화함수로 사용한 것으로 보아 10개의 class에 대한 확률 벡터를 출력할 것이다.

compile() 메소드를 사용해서 optimizer, loss, metrics을 설정한다.

### optimizer
optimizer로는 adam을 사용한다.

tf.keras.optimizers에서 지원하고 있는 optimizer를 사용할 수 있다.

### loss
loss(손실함수)로는 sparse_categorical_crossentropy를 사용한다.

tf.losses에서 제공하는 함수를 사용할 수 있다.

만약 output이 여러개라면 loss의 리스트나 딕셔너리를 넘겨줌으로써 각각 다른 loss를 사용할 수 있다.

loss 값은 각 loss들의 합에 의해서 최소화된다.

### metrics
metrics은 accuracy를 사용한다.

학습과 테스트 과정에서 평가로 사용할 metric의 list를 전달해줘야한다. 일반적으로 ['accuracy']를 사용한다.

`metrics={'output_a': 'accuracy', 'output_b': ['accuracy', 'mse']}`
이런식으로 사용하기도 한다.

그 외의 다양한 매개변수에 대한 정보는 
https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile
를 참고하자.


In [ ]:
model.fit(x_train, y_train, epochs=5)

model.evaluate(x_test, y_test, verbose=2)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 5s 78us/sample - loss: 0.2953 - acc: 0.9136
Epoch 2/5
60000/60000 [==============================] - 4s 75us/sample - loss: 0.1440 - acc: 0.9570
Epoch 3/5
60000/60000 [==============================] - 5s 80us/sample - loss: 0.1068 - acc: 0.9674
Epoch 4/5
60000/60000 [==============================] - 4s 74us/sample - loss: 0.0886 - acc: 0.9721
Epoch 5/5
60000/60000 [==============================] - 4s 74us/sample - loss: 0.0765 - acc: 0.9764
10000/10000 - 0s - loss: 0.0722 - acc: 0.9782


[0.07215387922241352, 0.9782]

# 모델 훈련과 평가

**이 예제에서는 5회 반복하여 train data 6만개를 학습한다.**

**평가 단계에서는 test 데이터 10000개를 사용했다.**

**loss값이 0.0722 이고 예측 정확도가 97.8%인 모델이 학습되었음을 확인할 수 있다.**



model.fit() - 고정된 epochs 수만큼 모델을 학습한다.
```python
fit(
    x=None,
    y=None,
    batch_size=None,
    epochs=1,
    verbose=1,
    callbacks=None,
    validation_split=0.0,
    validation_data=None,
    shuffle=True,
    class_weight=None,
    sample_weight=None,
    initial_epoch=0,
    steps_per_epoch=None,
    validation_steps=None,
    validation_freq=1,
    max_queue_size=10,
    workers=1,
    use_multiprocessing=False,
    **kwargs
)
```
x : input data(주로 numpy array, tf tensor)

y : target data(주로 numpy array, tf tensor)

batch_size : 사용하게 되면 디폴트 값 32. train data를 32개의 배치로 구분한다.

epochs = 모델을 학습할 횟수.

verbose = 학습과정을 어떻게 보여줄 것인지.
0 : 아무것도 표시 안함.
1 : progress bar
2 : 각 epoch마다 one line 표시

이 외의 매개변수는 API 확인
https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit

model.evaluate() - test 모드에서 모델의 지정해줬던 metrics 값과 loss 값을 반환한다.

```python
evaluate(
    x=None,
    y=None,
    batch_size=None,
    verbose=1,
    sample_weight=None,
    steps=None,
    callbacks=None,
    max_queue_size=10,
    workers=1,
    use_multiprocessing=False
)
```

fit()에서의 파라미터들과 동일한 기능을 한다.
